pkg/portfwd/forward_test.go

```diff
@@ -0,0 +1,148 @@
+// SPDX-FileCopyrightText: Copyright The Lima Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package portfwd
+
+import (
+	"context"
+	"errors"
+	"net"
+	"testing"
+	"time"
+
+	"gotest.tools/v3/assert"
+
+	"github.com/lima-vm/lima/v2/pkg/guestagent/api"
+	"github.com/lima-vm/lima/v2/pkg/limatype"
+)
+
+func TestForwarderIgnoreSkipsDynamicListeners(t *testing.T) {
+	dial := func(context.Context, string, string) (net.Conn, error) {
+		return nil, errors.New("unexpected dial")
+	}
+
+	guestIP := net.ParseIP("127.0.0.1")
+	hostIP := net.ParseIP("127.0.0.1")
+	guestRange := [2]int{1, 65535}
+	falseVal := false
+	rules := []limatype.PortForward{{
+		Proto:             limatype.ProtoAny,
+		GuestIPMustBeZero: &falseVal,
+		GuestIP:           guestIP,
+		GuestPortRange:    guestRange,
+		HostIP:            hostIP,
+		HostPortRange:     guestRange,
+	}}
+
+	tests := []struct {
+		name      string
+		protocol  string
+		ignoreTCP bool
+		ignoreUDP bool
+	}{
+		{name: "tcp ignored", protocol: "tcp", ignoreTCP: true},
+		{name: "udp ignored", protocol: "udp", ignoreUDP: true},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+		t.Run(tc.name, func(t *testing.T) {
+			preCheck := NewPortForwarder(rules, false, false)
+			defer preCheck.Close()
+			port := &api.IPPort{Protocol: tc.protocol, Ip: "127.0.0.1", Port: 23456}
+			local, _ := preCheck.forwardingAddresses(port)
+			assert.Assert(t, local != "", "test precondition failed: expected forwarding address for %s", tc.protocol)
+
+			fw := NewPortForwarder(rules, tc.ignoreTCP, tc.ignoreUDP)
+			fw.OnEvent(t.Context(), dial, &api.Event{AddedLocalPorts: []*api.IPPort{port}})
+			assert.Equal(t, 0, len(fw.closableListeners.listeners))
+			assert.Equal(t, 0, len(fw.closableListeners.udpListeners))
+			assert.NilError(t, fw.Close())
+		})
+	}
+}
+
+type nopListener struct{}
+
+func (nopListener) Accept() (net.Conn, error) { return nil, errors.New("accept not supported") }
+func (nopListener) Close() error              { return nil }
+func (nopListener) Addr() net.Addr            { return &net.TCPAddr{} }
+
+type nopPacketConn struct{}
+
+func (nopPacketConn) ReadFrom([]byte) (int, net.Addr, error) {
+	return 0, nil, errors.New("read not supported")
+}
+
+func (nopPacketConn) WriteTo([]byte, net.Addr) (int, error) {
+	return 0, errors.New("write not supported")
+}
+func (nopPacketConn) Close() error                     { return nil }
+func (nopPacketConn) LocalAddr() net.Addr              { return &net.UDPAddr{} }
+func (nopPacketConn) SetDeadline(time.Time) error      { return nil }
+func (nopPacketConn) SetReadDeadline(time.Time) error  { return nil }
+func (nopPacketConn) SetWriteDeadline(time.Time) error { return nil }
```

Comment:
This looks little bit too complicated. Complicated test mocks are likely to be broken and give incorrect results. Can we create a much simpler interface that is easy to test?

For example, the code handling events can use something like forwardPort() interface doing the right thing for the protocol. In this case we need to mock a single function.